rosetta/transactions/intent.go

```diff
@@ -0,0 +1,155 @@
+	if err != nil {
+		return nil, fmt.Errorf("invalid receiver account: %w", err)
+	}
+
+	// Validate the currencies specified for deposit and withdrawal.
+	send.Amount.Currency, err = p.validate.Currency(send.Amount.Currency)
+	if err != nil {
+		return nil, fmt.Errorf("invalid sender currency: %w", err)
+	}
+	receive.Amount.Currency, err = p.validate.Currency(receive.Amount.Currency)
+	if err != nil {
+		return nil, fmt.Errorf("invalid receiver currency: %w", err)
+	}
+
+	// Make sure that both the send and receive operations use the same currency.
+	// This is perhaps unnecessary at the moment since we only have a single currency.
+	if send.Amount.Currency != receive.Amount.Currency {
+		return nil, failure.InvalidIntent{
+			Sender:      send.AccountID.Address,
+			Receiver:    receive.AccountID.Address,
+			Description: failure.NewDescription("send and receive currencies do not match"),
+		}
+	}
+
+	// Parse value specified by the sender, after removing the negative sign prefix.
+	trimmed := strings.TrimPrefix(send.Amount.Value, "-")
+	sv, err := convert.ParseRosettaValue(trimmed)
+	if err != nil {
+		return nil, failure.InvalidIntent{
+			Sender:   send.AccountID.Address,
+			Receiver: receive.AccountID.Address,
+			Description: failure.NewDescription("could not parse withdrawal amount",
+				failure.WithString("withdrawal_amount", send.Amount.Value),
+				failure.WithErr(err),
+			),
+		}
+	}
+	// Parse value specified by the receiver.
+	rv, err := convert.ParseRosettaValue(receive.Amount.Value)
+	if err != nil {
+		return nil, failure.InvalidIntent{
+			Sender:   send.AccountID.Address,
+			Receiver: receive.AccountID.Address,
+			Description: failure.NewDescription("could not parse deposit amount",
+				failure.WithString("deposit_amount", receive.Amount.Value),
+				failure.WithErr(err),
+			),
+		}
+	}
+
+	// Check if the specified amounts match.
+	if sv != rv {
+		return nil, failure.InvalidIntent{
+			Sender:   send.AccountID.Address,
+			Receiver: receive.AccountID.Address,
+			Description: failure.NewDescription("deposit and withdrawal amounts do not match",
+				failure.WithString("deposit_amount", receive.Amount.Value),
+				failure.WithString("withdrawal_amount", send.Amount.Value),
+			),
+		}
+	}
+
+	if strings.ToUpper(send.Type) != dps.OperationTransfer ||
+		strings.ToUpper(receive.Type) != dps.OperationTransfer {
+
+		return nil, failure.InvalidIntent{
+			Sender:   send.AccountID.Address,
+			Receiver: receive.AccountID.Address,
+			Description: failure.NewDescription("only transfer operations are supported",
+				failure.WithString("deposit_type", receive.Type),
+				failure.WithString("withdrawal_type", send.Type),
+			),
+		}
+	}
+
+	intent := Intent{
+		From:     flow.HexToAddress(send.AccountID.Address),
+		To:       flow.HexToAddress(receive.AccountID.Address),
+		Amount:   sv,
+		Payer:    flow.HexToAddress(send.AccountID.Address),
+		Proposer: flow.HexToAddress(send.AccountID.Address),
```

Comment:
```suggestion
		From:     flow.HexToAddress(send.AccountID.Address),
		Payer:    flow.HexToAddress(send.AccountID.Address),
		Proposer: flow.HexToAddress(send.AccountID.Address),
		To:       flow.HexToAddress(receive.AccountID.Address),
		Amount:   sv,
```

Feels clearer to me like this, but it's a massive nitpick so feel free to disregard haha